### PR TITLE
Feat: Refactor & add logs to module upgrade

### PR DESCRIPTION
**Pull request by ga-devfront**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Refactor & add logs to module upgrade
| Type?             | improvement
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | --
| Sponsor company   | @PrestaShopCorp
| How to test?      | Then you this PR you can see more logs inside your console then modules are upgraded (localy or from addons).

### CODE REVIEW

File: classes/UpgradeTools/Module/ModuleDownloader.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        if ($this->moduleIsLocal) {
+            $this->downloadModuleFromLocalZip();
+        }
+
+        if (!$this->localModuleUsed) {
+            $this->downloadModuleFromAddons();
+        }
+
+        if (filesize($this->zipFullPath) <= 300) {
+            unlink($this->zipFullPath);
```
Comment (Quetzacoalt91) [inline classes/UpgradeTools/Module/ModuleDownloader.php:81]: We can move the removal of the file in the caller

```suggestion
```

File: classes/UpgradeTools/Module/ModuleUnzipper.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        if ($this->zipAction === null || $this->zipFullPath === null || $this->modulesPath === null || $this->moduleName === null) {
+            throw (new LogicException('Module unzip context is empty, please run setUnzipContext() first.'));
+        }
+
+        if (!$this->zipAction->extract($this->zipFullPath, $this->modulesPath)) {
+            throw (new UpgradeException($this->translator->trans('[WARNING] Error when trying to extract module %s.', [$this->moduleName])))->setSeverity(UpgradeException::SEVERITY_WARNING);
+        }
+        if (file_exists($this->zipFullPath)) {
+            unlink($this->zipFullPath);
+        }
```
Comment (Quetzacoalt91) [inline classes/UpgradeTools/Module/ModuleUnzipper.php:58]: We can move the removal of the file in the caller so it called in both scenarios (failure & success)

```suggestion
```

File: classes/Task/Upgrade/UpgradeModules.php [hunk 3]
```diff
@@ -76,11 +81,36 @@ public function run(): int
...
+                    $moduleMigration = new ModuleMigration($this->translator, $this->logger);
+                    $moduleMigration->setMigrationContext($module, $dbVersion);
+
+                    if (!$moduleMigration->needMigration()) {
+                        $this->logger->info($this->translator->trans('Module %s does not need to be migrated.', [$moduleInfos['name']]));
+                    } else {
+                        $moduleMigration->runMigration();
+                    }
+                    $this->logger->info($this->translator->trans('The module %s has been updated.', [$moduleInfos['name']]));
                 } catch (UpgradeException $e) {
```
Comment (Quetzacoalt91) [inline classes/Task/Upgrade/UpgradeModules.php:114]: Add after this catch block 

```php
finally {
    // Cleanup of module assets
    (new Filesystem)->remove([$zipFullPath]);
}

File: classes/UpgradeTools/Module/ModuleDownloader.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /**
+     * @throws UpgradeException
+     */
+    private function downloadModuleFromAddons(): void
+    {
+        $addonsUrl = extension_loaded('openssl')
+            ? 'https://api.addons.prestashop.com'
+            : 'http://api.addons.prestashop.com';
```
Comment (M0rgan01) [inline classes/UpgradeTools/Module/ModuleDownloader.php:109]: `api.addons.prestashop.com` could be a class constant

File: classes/UpgradeTools/Module/ModuleDownloader.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+use LogicException;
+use PrestaShop\Module\AutoUpgrade\Exceptions\UpgradeException;
+use PrestaShop\Module\AutoUpgrade\Log\Logger;
+use PrestaShop\Module\AutoUpgrade\Tools14;
+use PrestaShop\Module\AutoUpgrade\UpgradeTools\Translator;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+
+class ModuleDownloader
+{
```
Comment (M0rgan01) [inline classes/UpgradeTools/Module/ModuleDownloader.php:14]: Don't forget the license header

**Comment by Hlavtox** [comment]
@Quetzacoalt91 @ga-devfront Hi guys, one question/wish related to the topic:

When I was upgrading from 1.7.6 to 8.1.7, I met an annoying issue when all the modules that were in the zip were copied to the shop. This is both inconvinient and dangerous, because there could be uninstalled modules with security holes the merchant doesn't know about. Do you think you could improve this? Or maybe you already did in some previous PR?

ping @kpodemski

**Comment by Quetzacoalt91** [comment]
> @Quetzacoalt91 @ga-devfront Hi guys, one question/wish related to the topic:
> 
> When I was upgrading from 1.7.6 to 8.1.7, I met an annoying issue when all the modules that were in the zip were copied to the shop. This is both inconvinient and dangerous, because there could be uninstalled modules with security holes the merchant doesn't know about. Do you think you could improve this? Or maybe you already did in some previous PR?

This is a first step in the improvement of module management during the upgrade, where we take control of the whole process of download / unzip / migrate inside the module first.
There was another PR merged a month ago: https://github.com/PrestaShop/autoupgrade/pull/699 and there is more to expect. :+1:

**Review by M0rgan01 — APPROVED** [review]

File: classes/UpgradeTools/Module/ModuleMigration.php [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                throw (new UpgradeException($this->translator->trans('[WARNING] Method %s does not exist. Module %s disabled.', [$methodName, $moduleMigrationContext->getModuleName()])))->setSeverity(UpgradeException::SEVERITY_WARNING);
+            }
+
+            // @phpstan-ignore deadCode.unreachable (we ignore this error because the previous if can be true or false)
+            try {
+                if (!$methodName($moduleMigrationContext->getModuleInstance())) {
+                    throw (new UpgradeException($this->translator->trans('[WARNING] Migration failed while running the file %s. Module %s disabled.', [basename($migrationFilePath), $moduleMigrationContext->getModuleName()])))->setSeverity(UpgradeException::SEVERITY_WARNING);
+                }
+            } catch (UpgradeException $e) {
+                $moduleMigrationContext->getModuleInstance()->disable();
```
Comment (Quetzacoalt91) [inline classes/UpgradeTools/Module/ModuleMigration.php:128]: Unnecessary call to disable as the second catch will be triggered as well.

File: classes/UpgradeTools/Module/ModuleMigration.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                throw $e;
```
Comment (Quetzacoalt91) [inline classes/UpgradeTools/Module/ModuleMigration.php:129]: We use to have this specific catch block to disable the module with UpgradeExceptions.

If you disable in all cases, you can remove it.

```suggestion
```

**Review by M0rgan01 — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @ga-devfront 

Thank you for your PR, I tested it and it seems to works as you can see : 

image


Tested from : 
1.7.8.8 to 8.1.7 
8.0.5 to 8.1.7 
8.1.5 to 8.1.7 
8.1.7 to 9.0.0
8.1.5 to 9.0.0
8.0.5 to 9.0.0

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you

**Pull request merged**